interface/ceed-basis.c

```diff
@@ -691,37 +686,45 @@ int CeedHouseholderApplyQ(CeedScalar *mat_A, const CeedScalar *mat_Q, const Ceed
 /**
   @brief Return pseudoinverse of a matrix
 
-  @param[in]     ceed Ceed context for error handling
-  @param[in]     mat  Row-major matrix to be factorized in place
-  @param[in]     m    Number of rows
-  @param[in]     n    Number of columns
+  @param[in]     ceed      Ceed context for error handling
+  @param[in]     mat       Row-major matrix to be factorized in place
+  @param[in]     m         Number of rows
+  @param[in]     n         Number of columns
   @param[out]    mat_pinv  Row-major pseudoinverse matrix
 
   @return An error code: 0 - success, otherwise - failure
 
   @ref Utility
 **/
 int CeedMatrixPseudoinverse(Ceed ceed, CeedScalar *mat, CeedInt m, CeedInt n, CeedScalar *mat_pinv) {
-  CeedScalar *tau, *I;
+  CeedScalar *tau, *I, *mat_copy;
 
-  CeedCall(CeedCalloc(m * m, &I));
   CeedCall(CeedCalloc(m, &tau));
-  // -- QR Factorization, mat = Q R
-  CeedCall(CeedQRFactorization(ceed, mat, tau, m, n));
-  // -- mat_pinv = R_inv Q^T
-  for (CeedInt i = 0; i < m; i++) I[i * m + i] = 1.0;
-  // ---- Apply R_inv, mat_pinv = I R_inv
-  for (CeedInt i = 0; i < m; i++) {  // Row i
-    mat_pinv[m * i] = I[n * i] / mat[0];
-    for (CeedInt j = 1; j < n; j++) {  // Column j
-      mat_pinv[j + m * i] = I[j + n * i];
-      for (CeedInt k = 0; k < j; k++) mat_pinv[j + m * i] -= mat[j + n * k] * mat_pinv[k + m * i];
-      mat_pinv[j + m * i] /= mat[j + n * j];
+  CeedCall(CeedCalloc(n * n, &I));
+  CeedCall(CeedCalloc(m * n, &mat_copy));
+  memcpy(mat_copy, mat, m * n * sizeof mat[0]);
+
+  // QR Factorization, mat = Q R
+  CeedCall(CeedQRFactorization(ceed, mat_copy, tau, m, n));
+
+  // mat_pinv = R_inv Q^T
+  for (CeedInt i = 0; i < n; i++) I[i * n + i] = 1.0;
+  // -- Apply R_inv, mat_pinv = I R_inv
+  for (CeedInt i = 0; i < n; i++) {  // Row i
+    mat_pinv[n * i] = I[n * i] / mat_copy[0];
+    for (CeedInt j = 1; j < m; j++) {  // Column j
+      mat_pinv[j + n * i] = I[j + n * i];
+      for (CeedInt k = 0; k < j; k++) mat_pinv[j + n * i] -= mat_copy[j + m * k] * mat_pinv[k + n * i];
+      mat_pinv[j + n * i] /= mat_copy[j + n * j];
     }
   }
```

Comment:
@jeremylt the issue is in `j=1:m` loop. Right now with this structure apparently we have only issues on some qfunction tests with this message
```
/home/rezgar/libCEED/backends/memcheck/ceed-memcheck-qfunction.c:60 in CeedQFunctionApply_Memcheck(): QFunction output 0 entry 0 is NaN after restoring write-only access
```
and basis tests look fine, for example
```
build/t311-basis /cpu/self/ref/serial
(base) rezgar@rezgar:~/libCEED$ build/t311-basis /cpu/self/memcheck/serial
```
but actually `valgrind` shows a leak in `t311-basis`  like
```
==6684== Invalid read of size 8
==6684==    at 0x4869D59: CeedMatrixPseudoinverse (ceed-basis.c:718)
==6684==    by 0x486B578: CeedBasisGetCollocatedGrad (ceed-basis.c:332)
==6684==    by 0x4889957: CeedBasisCreateTensorH1_Ref (ceed-ref-basis.c:297)
==6684==    by 0x486A20E: CeedBasisCreateTensorH1 (ceed-basis.c:1018)
==6684==    by 0x4869F3A: CeedBasisCreateTensorH1 (ceed-basis.c:987)
==6684==    by 0x486D1A5: CeedBasisCreateTensorH1Lagrange (ceed-basis.c:1089)
==6684==    by 0x109469: main (t311-basis.c:39)
==6684==  Address 0x4f78c20 is 0 bytes after a block of size 96 alloc'd
==6684==    at 0x484DA83: calloc (in /usr/libexec/valgrind/vgpreload_memcheck-amd64-linux.so)
==6684==    by 0x486F056: CeedCallocArray (ceed.c:229)
==6684==    by 0x4869C1F: CeedMatrixPseudoinverse (ceed-basis.c:704)
==6684==    by 0x486B578: CeedBasisGetCollocatedGrad (ceed-basis.c:332)
==6684==    by 0x4889957: CeedBasisCreateTensorH1_Ref (ceed-ref-basis.c:297)
==6684==    by 0x486A20E: CeedBasisCreateTensorH1 (ceed-basis.c:1018)
==6684==    by 0x4869F3A: CeedBasisCreateTensorH1 (ceed-basis.c:987)
==6684==    by 0x486D1A5: CeedBasisCreateTensorH1Lagrange (ceed-basis.c:1089)
==6684==    by 0x109469: main (t311-basis.c:39)
```
and I printed the indices and values of `mat_copy`  and `mat_pinv` which shows we are out of range/size m x n and also have some nan/inf values. I don't know how with these values in `mat_pinv` after applying Q we have correct values for collocated grad.

Reply:
That's not quite a leak - a leak would be unfreed memory - but its definitely an issue. `mat_pinv[j + n * i]` would be assuming the matrix is NxN, so I agree with Valgrind that's a dimensions mismatch if N > M

Reply:
```
/var/lib/gitlab-runner/builds/N8GCsqus/0/libceed/libCEED/backends/memcheck/ceed-memcheck-qfunction.c:60 in CeedQFunctionApply_Memcheck(): QFunction output 0 entry 0 is NaN after restoring write-only access
```

this means that either the QFunction is not covering all qpts (which we know isn't the case) or the QFunction is getting NaNs in as input and then writing them out. If you run with  `CEED_DEBUG=1`, you should see an error about the write only access still having a NaN earlier in the OperatorApply, after `CeedBasisApply` for the input, I'd bet